hypervisor/arch/x86/guest/vm.c

```diff
@@ -130,6 +130,40 @@ static void create_prelaunched_vm_e820(struct acrn_vm *vm)
 	vm->e820_entry_num = E820_MAX_ENTRIES;
 	vm->e820_entries = (struct e820_entry *)ve820_entry;
 }
+
+/**
+ * @pre vm != NULL && vm_config != NULL
+ */
+static void create_ept_from_e820(struct acrn_vm *vm, const struct acrn_vm_config *vm_config)
+{
+	uint64_t base_hpa = vm_config->memory.start_hpa;
+	uint32_t i;
+
+	for (i = 0U; i < vm->e820_entry_num; i++) {
+		struct e820_entry *entry = &(vm->e820_entries[i]);
+
+		if (entry->length == 0UL) {
+			break;
+		}
+
+		/* Do EPT mapping for GPAs that are backed by physical memory */
+		if ((entry->type == E820_TYPE_RAM) && (entry->baseaddr >= (uint64_t)MEM_1M)) {
+			uint64_t base_gpa = entry->baseaddr;
+			uint64_t size = entry->length;
+
+			/* Merge memory under 1MB to lowmem so that it can be mapped with larger page size */
+			if (base_gpa == (uint64_t)MEM_1M) {
```

Comment:
This assumes that the 0-1M memory has a separate e820 entry, and the next e820 entry doesn't contain 1M-2M space. This is strange?

Reply:
It only assumes there is an entry starts from 1M memory. It's for pre-launched VM whose ve820.c is statically defined in .c code.

The code could be complicated if make it generic and without assumption.

Reply:
MPTable use e820 reserved region so that OS does not claim and use that memory.
Instead of this approach, we can run thru all e820 entries and assign EPT/HPA for type E820_TYPE_RAM in this function.
Then in mptable_build API, we would add an EPT region for MPTable.
Drawbacks with this approach is that First 1 MB of GPA will be fragmented and would be using 4K EPT pages, and also we need to maintain where HPA has ended after every ept_mr_add somewhere in acrn_vm struct.
But code looks cleaner and sticks to the basics of E820 type definition.

Reply:
> MPTable use e820 reserved region so that OS does not claim and use that memory.
> Instead of this approach, we can run thru all e820 entries and assign EPT/HPA for type E820_TYPE_RAM in this function.
> Then in mptable_build API, we would add an EPT region for MPTable.
> Drawbacks with this approach is that First 1 MB of GPA will be fragmented and would be using 4K EPT pages, and also we need to maintain where HPA has ended after every ept_mr_add somewhere in acrn_vm struct.
> But code looks cleaner and sticks to the basics of E820 type definition.

Sainath,
Thanks for the comments. one more thing, we need to leave 1MB hole in E820, otherwise, the lowmem part EPT will be 4K EPT pages.

Reply:
As discussed. the RAM section of e820 is better to start from 2MB rather than 1MB for simplicity... And simplify the code